Communications

Patrykz94 opened the pull request:
Adds the feature from issue #13 .

I think the functionality is basically ready now. I only added 2 types of data that you can request from PEGAS but that could always be expanded (just add entries to a lexicon), if you have ideas, let me know.

Let me know if you need anything changed or clarified.

Noiredd commented:
Thank you for the PR! I will review it in detail this weekend after I finish fixing the throttle system and a couple related things that I found.  
When I push my changes, I will ask you to squash your commits into a single one and rebase it on top of the then-latest version.

Noiredd commented:
Okay so the main thing that you want to achieve is shutting down the booster's engines and staging the it early - in this order. These are two separate things. For the comms, the idea is to have a listener that triggers when a message has been sent, and handles the message in some way. Then you propose a method of causing the stage to shutdown earlier than initially defined. There are a couple things that I do not fully understand at this moment, and a few that I would recommend doing in a different way. I'll briefly list them:
1. All event systems in PEGAS work on a principle of tiny triggers and handlers. That is, we have an initialization function (eg. `setSystemEvents`) that sets up the first trigger; the trigger body only sets the event occurence flag (here: `systemEventFlag`); *then* the main routine, having encountered the event flag, runs the handling function (`systemEventHandler`). This way we avoid executing event code at unpredictable moments, which would pose a risk of clogging the system (as an event might light up during the UPFG calculation, which we need to run fast). I would see the comm system working in the same way.
2. It would be good to make the `message` a kind of standardized structure, so that we can extend it in the future. I imagine it having a `type` key that would determine what is the message supposed to do (`request` or `command`, with potential of adding new kinds of messages) and `data` key holding requested variable names, names of commands to be executed, values etc. This way we could also pass function delegates into the system.  
I see this kind of solution as cleaner and less error-prone than checking whether the message contains a given key. What if a message contains more keys or is in some other way malformed? Let me know your thoughts.
3. For the early shutdown, you accomplish it by adding a new key to the `staging` sub-struct of a stage. If I read correctly, when a staging event occurs, you add a temporary key `SHUTDOWN` to `staging`, so that the throttle controller cut the engine when encountering it? But the `throttleControl` does not run until we're in the active guidance phase - how can this work? Also, why all the `IF event:HASKEY("SHUTDOWN") { event:REMOVE("SHUTDOWN"). }` lines? IMHO it would be simpler to modify the `stageEventHandler` so that it cut the throttle simply in the case when a stage has `shutdownRequired` - still that is a whole separate issue (than the comms system).

At this point I realize that it would be quite an effort to patch up and rebase this PR. If you want to keep working on this - let me know. Otherwise, I am willing to take lead on this - in this case I'd welcome any ideas that would make your life easier (commands you would like to be able to send to PEGAS, stuff you'd like to be able to request from it etc.).

Patrykz94 commented:
First of all, thanks for the feedback. I am working to fix those issues now but if you want to take over instead then I'm also fine with that.

1. I should have really made it more uniform with the rest of PEGAS (regarding events), I'm redoing this again, on the new version so I might update this PR tomorrow if I have some spare time.

2. Sounds good, doing it this way now.

3. Once I sent the message with new staging time (5 seconds in advance), PEGAS goes into UPFG mode so that's why `throttleControl` worked. As for `IF event:HASKEY("SHUTDOWN") { event:REMOVE("SHUTDOWN"). }` lines, I just put them after each `stage.` command so that as soon as the second stage is separated, `throttleControl` regains control over throttle. Thought that's not the prettiest way to do it. I'm trying to come up with a better way to do it, ideally without having to provide the new staging time with advance.

Also, would you like me to do communications and staging in separate PRs?

Noiredd commented:
Regarding point 3, do I understand correctly that if one *didn't* send the staging time update command, it wouldn't work? I'm a little confused by my own code now, mainly the `upfgConvergenceDelay`.

Let's keep this PR about what the title says: comms. If you want to continue to work on it - go ahead. Just a reminder: I pushed a few updates recently, so you should merge/rebase first to fix the conflicts. Also, do try to keep the number of commits minimum (a good idea is to squash before pushing, or alternatively a soft reset and reorganizing the changes into a few well organized commits - if you feel it would be logical to do so).  
In the meantime, I will take care of the shutdown thing.

Patrykz94 commented:
The only thing the staging time update command does is change the `upfgActivation` time to a new value and if `nextStageTime` is defined (which I just realized would never be defined at this point) it is also changed.
The `shutdownRequired` should work regardles of the message being sent or not.

And yes, I reset all changes and updated to newest version of pegas.

Patrykz94 commented:
I haven't finished this yet but here's a working prototype of the requesting part. I also added a temporary testing file.

Hope you don't mind that I got rid the old commits.

Function delegates can't be sent via message because they only work with serializable (lists, lexicons etc) and primitive data (string, scalars).

Also, for sending commands, I'm thinking of using a very similar way of having a lexicon with available commands in form of anonymous functions. Script would check if that command is available and if yes, execute it. Possibly sending a response that it has been executed successfully. What do you think about this?

Patrykz94 commented on kOS/pegas_util.ks:
```diff
@@ -415,6 +415,14 @@ FUNCTION setUserEvents {
+FUNCTION setComms {
+	WHEN NOT CORE:MESSAGES:EMPTY THEN {
+		SET commsEventFlag TO TRUE.
+		RETURN TRUE.				//	Preserves the trigger
```
Will need to change this part, if message isn't handled straight away (e.g. if in the middle of setting up UPFG when messages are not being checked), this trigger will fire up every tick.
Instead of preserving the trigger, I could maybe call `setComms()` again after handling and removing a message from queue.

Patrykz94 replied:
Done.

Patrykz94 commented on kOS/commsTest.ks:
```diff
@@ -0,0 +1,40 @@
+
+LOCAL requestData IS LEXICON(
+	"type", "request",
+	"data", "liftOffTime",
```
To request multiple things, a list of requested data can be provided, e.g. `"data", list("liftOffTime", "launchAzimuth"),`

Noiredd replied:
Speaking of it, we will need to come up with some documentation too, if this is to become a fully-formed feature.

Patrykz94 replied:
Done.

Noiredd commented on kOS/pegas_util.ks:
```diff
@@ -735,6 +743,56 @@ FUNCTION stageEventHandler {
+		}
+
+	} ELSE IF message["type"] = "command" {
+		//	TO BE CREATED
```
On your [comment](https://github.com/Noiredd/PEGAS/pull/14#issuecomment-337448288) - yes, I think we should do it the same way: list available commands (btw, I like the idea of putting delegates into the lexicon) and have the handler call whatever value it finds under the received key.  
About what the command should return - I think checking whether a command executed properly is not very realistic (how do we make sure of that?). Instead, I would see each delegate returning some value (from simple true/false to more detailed information) which we would then pack into the reply, and have it up to the sender to figure out what to do with them. This reiterates the need for documentation.

On the other hand, how do you feel about the sender being able (or not) to send multiple commands (similar to multiple requests)?

Patrykz94 replied:
I guess multiple commands could be useful in some situations so we could add that.
If we were to have only one command per message, the message could include:
`... "data", LIST("engineShutdown", "engine_1_tag", "engine_2_tag") ...`
So a list with first item being the command name, and the remaining items being parameters.
For multiple commands per message, it could be done like this:
`... "data", LEXICON("engineShutdown", LIST("engine_1_tag", "engine_2_tag")) ...`
So a lexicon with command name as key and list of parameters as values. This could have multiple entries.

I'd be more inclined towards the second option with multiple commands, what do you think?

Noiredd replied:
Are lexicons order-preserving? I'm used to Python's `dict`s and C++'s `map`s, which are not - and if we allow the user to send multiple commands, we should make sure that they will be executed in the exact order the user passes them. For this reason I would lean towards lists here. The `data` key would contain a list of lists in the first format:  
`"data", LIST(LIST("engineShutdown", "tag_1", "tag_2"), LIST("setVar", "asd", 130), ...)`  
For ease of use we should do like in the request message, allow omitting the outermost list and have the handler ensure the format adheres to the standard before running the loop.

Patrykz94 replied:
I don't think I ever noticed a lexicon changing order but I made it the way you suggested.

Patrykz94 commented on kOS/pegas_util.ks:
```diff
@@ -735,6 +743,56 @@ FUNCTION stageEventHandler {
+		//	Create a response to send back
+		LOCAL response IS LEXICON("type", "response").
+
+		//	Iterate through requested data and add to response
```
Yeah, that makes sense. In case there is something wrong, should we return an actual error message (e.g. "Requested data not found") or just a simple "ERROR"?
Error message would be more descriptive for a human, but a simple "ERROR" would be easier to handle for a script.
Which one would you prefer or do you have any other idea?

Noiredd replied:
`"ERROR (Requested data not found)"` is equally easy to parse for the program that made the request (`RESPONSE:ENDSWITH("ERROR")` or something like that) and the human who might read the message on screen.

Patrykz94 replied:
Yeah, I forgot this is possible. Done now.

Patrykz94 commented:
I think it's pretty close to being ready for another review. I'll try to get documentation and a few commands done today and then I'll test everything tomorrow.

The commands that I'm adding now are `upfgTimeUpdate`, `engineShutdown` and `engineThrottle` (I'll make them locked to atmospheric ascent only). If you have any other ones, let me know. New commands can easily be added by creating a new entry in `availableCommands` lexicon (same goes for requesting data).

Patrykz94 commented:
Ready for a review 2.0.
I'm adding a testing file in case you want to use it.
[commsTest.zip](https://github.com/Noiredd/PEGAS/files/1442845/commsTest.zip)

Noiredd commented:
It looks very good! I see a couple minor things that I'd look deeper into and perhaps change, but I am almost ready to merge this as is.

There's just one thing I want to discuss first: how would you feel about making a whole separate module out of this - `pegas_comm.ks`? Event handler really grew, and I had this thought that maybe if we packed it into its own file we'd have more freedom in splitting this into atomic functions - for easier maintenance and extensibility. I imagined moving the message format fixing code to a dedicated function ("`verifyMessageFormat`"?) and the available command lexicon to a global variable (instead of building it every time the handler runs), built by taking delegates to named functions. Adding commands would then become as easy as writing a new function and putting a delegate to it to the lexicon. We could arrange the module for example like this:
* `setComms()`
* `commsEventHandler()`
* `verifyMessageFormat()`
* commands as separate, named functions, like `command_setThrottle`, `command_setUpfgTime` etc,
* `GLOBAL availableCommands IS LEXICON("setThrottle", command_setThrottle@, ...).`

You've put a great deal of time and effort into this PR and I greatly appreciate it!

Patrykz94 commented:
Good idea, I'll work on this now.

Patrykz94 commented:
@Noiredd Pretty much ready, would you also like me to make `availableData` a global variable like `availableCommands`?

Noiredd commented:
Can this even be done? Commands can be stored as function delegates, but for variables, we need to read their values at the moment they were requested.  
At second thought however, I can imagine something like "`readAvailableData`" which takes name of the variable as parameter and returns its value or an error string. If you feel this approach would be better - go for it :)

Patrykz94 commented:
Never mind, forgot how variables work, lol :)

I'll push the updated version if a few minutes.

Patrykz94 commented:
Ok that's done.

I also added some extra checks before sending a response (if "sender" contained incorrect CPU name or if the sender CPU separated before the response, it would crash PEGAS).

With the `"engineShutdown"` command, this actually acheives everything I needed from my original issue #13. Although I think a `"shutdownRequired"` option could probably still be useful if you decided to do it.

Noiredd commented:
Great job - a really high quality contribution! I toyed around with it a bit and it looks perfect. As soon as I push my own commits, I will merge this as well. Thank you, @Patrykz94!

In the meantime, I actually coded the `"shutdownRequired"` option. Its advantage over purely using the comms is that you don't need to tag the engines (if all you want to do is shutdown the entire stage).
